File: bcmodel-rs/examples/load_model.ts
/**
 * Load a .bcmodel file in Node.js using the TypeScript wrapper.
 *
 * Prerequisites:
 *   cd bcmodel-rs/bcmodel-wasm
 *   wasm-pack build --target web
 *
 * Usage:
 *   npx tsx load_model.ts ../../meshnet/model5_gw_ae/model.bcmodel
 */
import { readFileSync } from "fs";
import { initBcmodelSync, loadFromBuffer, computeTensorStats } from "../bcmodel-ts/src/index.js";

// --- Initialize WASM ---
const wasmBytes = readFileSync(
  new URL("../bcmodel-wasm/pkg/bcmodel_wasm_bg.wasm", import.meta.url)
);
initBcmodelSync(wasmBytes);

// --- Parse CLI args ---
const modelPath = process.argv[2];
if (!modelPath) {
  console.error("Usage: npx tsx load_model.ts <path/to/model.bcmodel>");
  process.exit(1);
}

// --- Load model ---
const data = new Uint8Array(readFileSync(modelPath));
const model = loadFromBuffer(data);

// --- Model info ---
console.log(`Name:         ${model.name}`);
console.log(`Type:         ${model.modelType}`);
console.log(`Input shape:  [${model.inputShape}]`);
console.log(`Classes:      ${model.numClasses}`);
console.log(`Graph nodes:  ${model.graphLength}`);
console.log(`Parameters:   ${model.totalParams.toLocaleString()}`);
console.log(`Weight size:  ${(model.weightSizeBytes / 1024).toFixed(1)} KB`);
console.log();

// --- Architecture graph ---
console.log("Architecture:");
for (const node of model.graph) {
  const inputs =
    node.inputs.length > 0
      ? ` <- ${node.inputs.join(", ")}`
      : " (input)";
  console.log(`  ${node.id.padEnd(20)} ${node.op.padEnd(15)}${inputs}`);
}
console.log();

// --- Tensors ---
console.log("Tensors:");
for (const name of model.tensorNames()) {
  const shape = model.tensorShape(name);
  const data = model.tensorData(name);
  if (data && shape) {
    const stats = computeTensorStats(name, data, shape);
    console.log(
      `  ${name.padEnd(30)} shape=[${shape}]  min=${stats.min.toFixed(4)}  max=${stats.max.toFixed(4)}`
    );
  }
}
console.log();

// --- Labels ---
const labels = model.labels;
if (labels && labels.length > 0) {
  console.log("Labels:");
  for (const label of labels) {
    const [r, g, b] = label.color;
    console.log(
      `  [${String(label.index).padStart(3)}] ${label.name.padEnd(20)} rgb(${r}, ${g}, ${b})`
    );
  }
  console.log();
}

// --- Inference config ---
const config = model.inferenceConfig;
if (config) {
  console.log("Inference config:");
  for (const [key, value] of Object.entries(config)) {
    console.log(`  ${key}: ${value}`);
  }
}

// --- Cleanup ---
model.free();

File: bcmodel-rs/examples/load_model.py
#!/usr/bin/env python3
"""Load a .bcmodel file and print model summary, tensors, and labels.

Prerequisites:
    cd bcmodel-rs/bcmodel-py
    python -m venv .venv && source .venv/bin/activate
    pip install numpy maturin
    maturin develop

Usage:
    python load_model.py ../../meshnet/model5_gw_ae/model.bcmodel
"""
import sys
from pathlib import Path

import bcmodel


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <path/to/model.bcmodel>")
        sys.exit(1)

    path = Path(sys.argv[1])
    if not path.exists():
        print(f"File not found: {path}")
        sys.exit(1)

    # --- Load model ---
    model = bcmodel.BcmodelFile.load(str(path))
    print(model)
    print()

    # --- Model info ---
    print(f"Name:         {model.name}")
    print(f"Type:         {model.model_type}")
    print(f"Description:  {model.description}")
    print(f"Input shape:  {model.input_shape}")
    print(f"Classes:      {model.num_classes}")
    print(f"Graph nodes:  {model.graph_length}")
    print(f"Parameters:   {model.total_params:,}")
    print(f"Weight size:  {model.weight_size_bytes / 1024:.1f} KB")
    print()

    # --- Architecture graph ---
    print("Architecture:")
    for node in model.graph():
        inputs = " <- " + ", ".join(node["inputs"]) if node["inputs"] else " (input)"
        print(f"  {node['id']:20s} {node['op']:15s}{inputs}")
    print()

    # --- Tensors ---
    print("Tensors:")
    for name in model.tensor_names():
        shape = model.tensor_shape(name)
        tensor = model.tensor(name)
        print(f"  {name:30s} shape={shape}  min={tensor.min():.4f}  max={tensor.max():.4f}")
    print()

    # --- Labels ---
    labels = model.labels()
    if labels:
        print("Labels:")
        for label in labels:
            r, g, b = label["color"]
            print(f"  [{label['index']:3d}] {label['name']:20s} rgb({r}, {g}, {b})")
    print()

    # --- Inference config ---
    config = model.inference_config()
    if config:
        print("Inference config:")
        for key, value in config.items():
            print(f"  {key}: {value}")

    # --- Compatibility API (drop-in replacement for load_bcmodel.py) ---
    print("\n--- Compatibility API ---")
    header, tensors = bcmodel.load_bcmodel(str(path))
    print(f"Version:  {header['bcmodel_version']}")
    print(f"Tensors:  {len(tensors)} arrays")
    first = next(iter(tensors))
    print(f"Example:  {first} -> shape={tensors[first].shape}, dtype={tensors[first].dtype}")


if __name__ == "__main__":
    main()
